Redesign mithril client api (#1332)
input-output-hk/mithril: docs/website/root/manual/developer-docs/nodes/mithril-client-library.md

Reviewer: You should use the following case in all the documentation:
- **Mithril client library**
- **Mithril network**

Mithril Client Library can be used by Rust developers to use the Mithril Network in their applications.

> It handles the different types that can be queried to a mithril aggregator:

Reviewer: ```suggestion
It is responsible for handling the different types of data certified by Mithril, and available through a Mithril aggregator:
```

## Installation

:::caution

> Mithril Client library has not yet been published on [crates.io](https://crates.io/), so you won't be able to follow the procedure below. The crate will be published soon.

Reviewer: Can you add an alternative option with a git path in the dependency configuration in the `Cargo.toml` file?

:::

> In your project, use `cargo` to add [mithril_client](https://crates.io/crates/mithril_client) crate as a dependency:

Reviewer: ```suggestion
In your project, use `cargo` to add [mithril-client](https://crates.io/crates/mithril-client) crate as a dependency:
```

```bash

> cargo add mithril_client

Reviewer: It should be `mithril-client`.

```

> It will add the latest version of the Mithril client library.

Reviewer: This sentence can be removed IMO

> Mithril client is an asynchronous library, you will need a runtime to execute your futures. We recommend to use the crate [tokio](https://crates.io/crates/tokio), as the library has been tested with it.

Reviewer: Can you add this in a tip or info block?

```rust title="/src/main.rs"
use mithril_client::{ClientBuilder, MessageBuilder};
use std::path::Path;

#[tokio::main]
async fn main() -> mithril_client::MithrilResult<()> {
    let client = ClientBuilder::aggregator("YOUR_AGGREGATOR_ENDPOINT", "YOUR_GENESIS_VERIFICATION_KEY").build()?;
    
    let snapshots = client.snapshot().list().await?;
    
    let last_digest = snapshots.first().unwrap().digest.as_ref();
    let snapshot = client.snapshot().get(last_digest).await?.unwrap();
    
    let certificate = client
        .certificate()
        .verify_chain(&snapshot.certificate_hash)
        .await?;
    
    // Note: the directory must already exist, and the user running this code must have read/write access to it.
    let target_directory = Path::new("YOUR_TARGET_DIRECTORY");
    client
        .snapshot()
        .download_unpack(&snapshot, target_directory)
        .await?;
    
    let message = MessageBuilder::new()
        .compute_snapshot_message(&certificate, target_directory)
        .await?;
    assert!(certificate.match_message(&message));
    
    Ok(())
}
```

:::info

> Snapshot download and Certificate chain validation can take quite some time even with a fast computer and network. For those a feedback mechanism is available, more details on it in the [feedback submodule](https://mithril.network/rust-doc/mithril_client/feedback/index.html).

Reviewer: ```suggestion
Snapshot download and certificate chain validation can take quite some time even with a fast computer and network. We have implemented a feedback mechanism for them, more details on it are available in the [feedback sub-module](https://mithril.network/rust-doc/mithril_client/feedback/index.html).
```

> An example of implementation with the crate [indicatif](https://crates.io/crates/indicatif) is available in the [Mithril repository](https://github.com/input-output-hk/mithril/tree/main/mithril-client/examples/snapshot_list_get_show_download_verify.rs).

Reviewer: Could you add a run command for the example?

:::

> Here is an example of the code for the release-preprod network:

Reviewer: ```suggestion
Here is a working example of the code using the configuration parameters of the `release-preprod` network:
```

:::caution

Mithril client library has not yet been published on [crates.io](https://crates.io/), so you won't be able to follow the procedure below. The crate will be published soon.

For now, you can experiment it by adding the dependency manually in the Cargo.toml of your project:

```toml title="/Cargo.toml"

> mithril-client = { git = "https://github.com/input-output-hk/mithril.git" }

Reviewer: Do we need to specify a path to the `mithril-client` crate?

Reviewer 2: No, it just needs this line to be able to use the import of the library in a project.

> If the goal is just to use the existing certificates, it is easier to use the `Client` structure:

Author: This is from the previous api doc that will be soon updated :).

> If the goal is just to use the existing certificates, it is easier to use the `Client` structure:

Author: The docusaurus has been updated ! Please review the changes :pray: .

:::info

Snapshot download and certificate chain validation can take quite some time even with a fast computer and network. We have implemented a feedback mechanism for them, more details on it are available in the [feedback sub-module](https://mithril.network/rust-doc/mithril_client/feedback/index.html).

An example of implementation with the crate [indicatif](https://crates.io/crates/indicatif) is available in the [Mithril repository](https://github.com/input-output-hk/mithril/tree/main/mithril-client/examples/snapshot_list_get_show_download_verify.rs). To run it, execute the following command:

```bash

> cargo run --example snapshot_list_get_show_download_verify 

Reviewer: Can you also add the `cd examples` command?

Reviewer 2: It's not needed, it can be run anywhere in the local repo 😉.
